enumerable.rb

```diff
@@ -0,0 +1,115 @@
+module Enumerable
+  def my_each
+    to_a.length.times { |i| yield self[i] }
+    self
+  end
+
+  def my_each_with_index
+    to_a.my_each { |each| yield each, index(each) }
+    self
+  end
+
+  def my_select
+    array = []
+    to_a.my_each { |each| array.push(each) if yield each }
+    array
+  end
+
+  def my_all?(parameter = nil)
+    to_a.my_each do |each|
+      if block_given?
+        return false unless yield each
+      elsif parameter.instance_of? Class
+        return false unless each.is_a? parameter
+      else
+        return false unless each
+        return false unless parameter === each
+      end
+    end
+    true
+  end
+
+  def my_any?(parameter = nil)
+    to_a.my_each do |each|
+      if block_given?
+        return true if yield each
+      elsif parameter.instance_of? Class
+        return true if each.is_a? parameter
+      else
+        return true if each == true
+        return true if parameter === each
+      end
+    end
+    false
+  end
+
+  def my_none?(parameter = nil)
+    to_a.my_each do |each|
+      if block_given?
+        return false if yield each
+      elsif parameter.instance_of? Class
+        return false unless each.is_a? parameter
+      else
+        return false if each == true
+        return false if none_nil?(parameter, each)
+      end
+    end
+    true
+  end
+
+  def none_nil?(parameter = nil, each = nil)
+    return true if !parameter.nil? && parameter === each
+
+    false
+  end
+
+  def my_count(parameter = nil, &block)
+    if block_given?
+      counter = to_a.my_select(&block)
+      counter.length
+    elsif parameter.nil?
+      to_a.length
+    else
+      counter = to_a.my_select { |each| each == parameter }
+      counter.length
+    end
+  end
+
+  def my_map(proc = nil)
+    return to_enum unless block_given? || !proc.nil?
+
+    array = []
+    if proc.nil?
+      to_a.my_each { |each| array.push(yield each) }
+    else
+      to_a.my_each { |_each| array.push(proc.call(to_a[i])) }
+    end
+    array
+  end
```

Comment:
## my_map
- [ ] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
array.my_map(my_proc) { |num| num < 10 }
```
got: NameError:
        undefined local variable or method `i' for #<Array:0x0000557be3bc2500>